### PR TITLE
Miri: detect wrong vtables in wide pointers

### LINKED ISSUE #3497: Wide pointer metadata validity: check that the vtable is for the right trait
With https://github.com/rust-lang/rust/pull/123572, rustc started assuming for UB purposes that a vtable pointer points to a vtabe that's sufficiently big for the given trait. Miri currently isn't able to check that UB.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/miri/issues/3497.
Needed to catch the UB that https://github.com/rust-lang/rust/pull/123572 will start exploiting.

r? @oli-obk

### CODE REVIEW

File: tests/ui/cast/cast-rfc0401-vtable-kinds.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ impl Foo<u32> for u32 { fn foo(&self, _: u32) -> u32 { self+43 } }
 impl Bar for () {}
 
 unsafe fn round_trip_and_call<'a>(t: *const (dyn Foo<u32>+'a)) -> u32 {
-    let foo_e : *const dyn Foo<u16> = t as *const _;
+    let foo_e : *const dyn Foo<u32> = t as *const _;
```
Comment (RalfJung) [inline tests/ui/cast/cast-rfc0401-vtable-kinds.rs:20]: Not sure if this intentionally changed the type or if it was intended to only cast away the lifetime. But we probably shouldn't have ui tests with UB. ;)

**Comment by GuillaumeGomez** [comment]
I think it detected new memory bugs which need to be fixed before it can be merged (as you can see in https://github.com/rust-lang/rust/pull/124264 CI output).

@bors r-

**Comment by RalfJung** [comment]
That looks to be just another case of https://github.com/rust-lang/rust/issues/123583.

@bors r=oli-obk

**Pull request merged**